Wrote blog post describing how to use Pulumi to create a Python AWS application with Flask and Redis (#3924)
pulumi/docs: content/blog/creating-a-python-aws-application-using-flask-and-redis/index.md

---
title: "Creating a Python AWS Application Using Flask and Redis"
date: 2020-08-11T15:06:26-07:00
meta_desc: A tutorial on how to create a Python AWS application using Flask, Redis, and Pulumi.
meta_image: meta.png
authors: ["vova-ivanov"]
tags: ["aws", "python", "ecs", "docker"]
---

*Meet Vova Ivanov---one of the Pulumi summer interns. He'll be writing about his experiences learning Pulumi while modernizing a web app and its underlying infrastructure.*

<!--more-->

I've recently started developing with Pulumi, and have begun to explore its inner workings and mechanisms. I decided to construct a production-level application and to document each step that I take and my progress as I go along.

This blog post recreates the existing [Typescript voting app example](https://www.pulumi.com/docs/tutorials/aws/aws-ts-voting-app/) step by step in Python with Flask as the frontend and Redis as the backend. In future blog posts, we will explore how to change the front and backends, how to upgrade the app with additional AWS services, and migrating from one cloud provider to another.

> The first few lines of our new `__main__.py` file indicate which libraries to import and describe a pair of configuration options used by the application

Reviewer: where does this `__main__.py` come from - there's nothing mentioned in here about the `pulumi project` structure

Author: Updated the paragraph to give a quick description of the project, and what the `__main__.py` file means to it.

            "protocol": "tcp"
        }],
        "environment": [
            { "name": "REDIS", "value": args[1]["host"] },
            { "name": "REDIS_PORT", "value": str(args[1]["port"]) },
            { "name": "REDIS_PWD", "value": redis_password },
        ],
    }])))
```

The service for our Flask frontend is launched. It will now receive requests from the internet,
process them, and change the Redis database accordingly.

```python
flask_service = aws.ecs.Service("flask-service",
	cluster=app_cluster.arn,
    desired_count=1,
    launch_type="FARGATE",
    task_definition=flask_task_definition.arn,
    wait_for_steady_state=False,
    network_configuration={
		"assign_public_ip": "true",
		"subnets": [app_vpc_subnet.id],
		"security_groups": [app_security_group.id]
	},
    load_balancers=[{
		"target_group_arn": flask_targetgroup.arn,
		"container_name": "flask-container",
		"container_port": 80,
	}],
    opts=pulumi.ResourceOptions(depends_on=[flask_listener]),
)
```

To connect to our application, we simply export the DNS name of the Flask balancer, and
open it in a browser window.

```python
pulumi.export("app-url", flask_balancer.dns_name)
```

> In this example, I showed how easy it is to convert infrastructure code defined in TypeScript to one defined in Python. Teams often choose to standardize on one language and Pulumi expands that choice, allowing them to standardize on one language for the frontend, the backend, and the infrastructure.

Reviewer 2: ```suggestion
In this example, I showed how straightforward it is to convert infrastructure code defined in TypeScript to one defined in Python. Teams often choose to standardize on one language and Pulumi expands that choice, allowing them to standardize on one language for the frontend, the backend, and the infrastructure.
```

> Next week, I'll show how to further modernize the application by shifting the frontend from Flask to Django, and swapping out the database from Redis to Amazon RDS---using Pulumi to stand up new resources as I go.

Reviewer 2: ```suggestion
Next week, I'll show how to modernize the application by changing the frontend from Flask to Django, and swapping out the database from Redis to Amazon RDS---using Pulumi to stand up new resources as I go.
```

This blog post recreates the existing [Typescript voting app example](https://www.pulumi.com/docs/tutorials/aws/aws-ts-voting-app/) step by step in Python with Flask as the frontend and Redis as the backend. In future blog posts, we will explore how to change the front and backends, how to upgrade the app with additional AWS services, and migrating from one cloud provider to another.

> The first step is to create a new directory and initialize a Pulumi project with `pulumi new aws-python`. Since we won't be modifying the application's frontend, we simply copy the `frontend/` folder from the example into our directory.

Reviewer 2: ```suggestion
The first step is to clone the [Typescript voting app example](https://www.pulumi.com/docs/tutorials/aws/aws-ts-voting-app/). We'll use a sparse clone to copy only the `aws-ts-voting-app` directory.

```bash
$ mkdir examples && cd examples
$ git init
$ git remote add origin -f https://github.com/pulumi/examples/
$ git config core.sparseCheckout true
$ echo aws-ts-voting-app >> .git/info/sparse-checkout
$ git pull origin master
``

The next step is to create a new directory and initialize a Pulumi project with `pulumi new aws-python`. 

```bash
$ mkdir aws-py-voting-app && cd aws-py-voting-app
$ pulumi new aws-python
```

Since we won't be modifying the application's frontend, we simply copy the `frontend/` folder from the example into our directory.

```bash
$ cp -r aws-ts-voting-app/frontend/ aws-py-voting-app/frontend
```
```

This blog post recreates the existing [Typescript voting app example](https://www.pulumi.com/docs/tutorials/aws/aws-ts-voting-app/) step by step in Python with Flask as the frontend and Redis as the backend. In future blog posts, we will explore how to change the front and backends, how to upgrade the app with additional AWS services, and migrating from one cloud provider to another.

> The project structure of our new Python voting app will be similar to the Typescript example, and the biggest difference will be the code describing the cloud infrastructure. Instead of an `index.ts` file, Pulumi will be given a new `__main__.py` file. The first few lines of the file indicate which libraries to import and describe a pair of configuration options used by the application.

Reviewer 2: ```suggestion
The project structure of our new Python voting app is similar to the Typescript example and the biggest difference is the code describing the cloud infrastructure. Instead of an `index.ts` file, the project uses `__main__.py` file. The first few lines of the file indicate which libraries to import and describe a pair of configuration options used by the application.
```

Reviewer: I think we should just talk about `pulumi new` here. The `__main__.py` isn't the only difference. The Pulumi.yaml has a `runtime: python` and we also need a requirements.txt file for the dependencies

If we scaffold a `pulumi new aws-python` app, we have all the components we need and it will help the users be able to recreate easily

$0.02

Author: Added a small tutorial on how to create the project, and incorporated Sophia's edits.
